core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala

```diff
@@ -184,10 +184,44 @@ class SnappySessionState(snappySession: SnappySession)
     }
   }
 
-  case class AnalyzeChildQuery(sparkSession: SparkSession) extends Rule[LogicalPlan] {
+  case class AnalyzeMutableOperations(sparkSession: SparkSession,
+      analyzer: Analyzer) extends Rule[LogicalPlan] {
+
+    private def getKeyAttributes(table: LogicalPlan,
+        child: LogicalPlan, plan: LogicalPlan): Seq[NamedExpression] = {
+      val keyColumns = table.collectFirst {
+        case LogicalRelation(mutable: MutableRelation, _, _) =>
+          val ks = mutable.getKeyColumns
+          if (ks.isEmpty) throw new AnalysisException(
+            s"Empty key columns for update/delete on $mutable")
+          ks
+      }.getOrElse(throw new AnalysisException(
```

Comment:
Can you please elaborate the usage of keyColumns in the scheme of things. Is it that we would support bulk update & delete only for tables having a primary key? Or I am missing something?